File: 03_Stack/largestRectangleInHistogram.js
// Function to find the largest rectangle area in a histogram using divide-and-conquer
var largestRectangleArea = function(heights, left = 0, right = (heights.length - 1)) {
    // Check if the current range is a base case (empty or single element)
    const isBaseCase = right < left;
    if (isBaseCase) return 0;

    // Call the divide-and-conquer helper function
    return divideAndConquer(heights, left, right);
}

// Helper function for divide-and-conquer approach
const divideAndConquer = (heights, left, right, min = left) => {
    // Iterate through the current range to find the index of the minimum height
    for (let i = left; i <= right; i++) {
        // Check if the current height is less than the height at the current minimum index
        const isMinGreater = heights[i] < heights[min];
        if (!isMinGreater) continue;

        // Update the index of the minimum height
        min = i;
    }

    // Calculate the width of the current range
    const window = (right - left) + 1;

    // Calculate the area of the rectangle formed by the minimum height in the current range
    const area = heights[min] * window;

    // Recursively calculate the area for the left and right subranges
    const leftArea = largestRectangleArea(heights, (min + 1), right);
    const rightArea = largestRectangleArea(heights, left, (min - 1));

    // Return the maximum of the area in the current range and the areas from the left and right subranges
    return Math.max(area, leftArea, rightArea);
}

// implementation
console.log("Leetcode 84: Largest Rectangle in Histogram\n");

// Example 1
const example1 = [2, 1, 5, 6, 2, 3];
console.log(`Example 1: heights = [${example1}]`);
console.log("Output:", largestRectangleArea(example1));

// Example 2
const example2 = [2, 4];
console.log(`Example 2: heights = [${example2}]`);
console.log("Output:", largestRectangleArea(example2));

File: 03_Stack/largestRectangleInHistogram.go
package main

import "fmt"

type StackValue struct {
	index  int // Index of the bar in the histogram
	height int // Height of the bar
}

func largestRectangleArea(heights []int) int {
	stack := []StackValue{} // Stack to track indices and heights of bars
	maxArea := 0            // Initialize maximum area
	var start int           // Variable to store the starting index of the rectangle

	// Iterate through each bar in the histogram
	for i, h := range heights {
		start = i // Update start index for the current bar
		// While stack is not empty and the height of the current bar is less than the height of the bar at the top of the stack
		for len(stack) != 0 && stack[len(stack)-1].height > h {
			// Pop the top bar from the stack
			index, height := stack[len(stack)-1].index, stack[len(stack)-1].height
			stack = stack[0 : len(stack)-1] // Pop operation
			// Calculate the area formed by the popped bar
			maxArea = max(maxArea, height*(i-index))
			start = index // Update start index for the rectangle
		}
		stack = append(stack, StackValue{start, h}) // Push the current bar to the stack
	}

	// Process any remaining bars in the stack
	for _, h := range stack {
		// Calculate the area formed by the remaining bars
		maxArea = max(maxArea, h.height*(len(heights)-h.index))
	}
	return maxArea // Return the maximum area
}

// Function to find the maximum of two integers
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// implementation
func main() {
	fmt.Println("Leetcode 84: Largest Rectangle in Histogram\n")

	// Example 1
	example1 := []int{2, 1, 5, 6, 2, 3}
	fmt.Printf("Example 1: heights = %v\n", example1)
	fmt.Println("Output:", largestRectangleArea(example1))

	// Example 2
	example2 := []int{2, 4}
	fmt.Printf("Example 2: heights = %v\n", example2)
	fmt.Println("Output:", largestRectangleArea(example2))
}
